### ISSUE TITLE
Live map caches might be missing on low zoom

### ISSUE THREAD

**Issue by Lineflyer** [cgeo/cgeo#8260]
This is a follow up issue for a regression caused by the fix for #8225:

While testing the interim solution we made, I ran into situations where the live map was no longer populated with caches at some areas. I do not yet have a clear sequence or description of the problem, but let me try to describe, what I think is happening:

If you zoom out the map on a region with high cache density, it will happen that we reach the maximum amount of caches returned from the server and this ony fills part of the total viewport requested (You will then typically see a circle filled with e.g. 500 caches in the center and empty space around).
If you now zoom in again staying within the area of the previously requested viewport but not within the "circle" of loaded caches, you will see a map without caches and those are not requested from the server as we consider this viewport already loaded.

Maybe we need to dig deeper, if it is that simple.
And we should really carefully analyze how to solve such kind of situation without again sending to many requests. One part of it could be to analyze the area covered by the caches loaded in the request and set this into relation with the total viewport. But this alone will not solve our problem as loading more caches will potentially cause other caches being thrown out of our internal cache cache.

I personally do not consider this a high priority problem as such, as the live map was never designed to visualize thousands of caches at once. So better stick with this regression than find a quick fix by increasing server load.

**Comment by Lineflyer** [comment]
There are quite some users on support channels which seem to run into this problem ("When I look on city A and then scroll to city B it is not showing any or at least not all caches there. I need to deactivate and activate the live mode to get the map updated").

Any idea about an improvement we could include into the next release?

**Comment by moving-bits** [comment]
Current caching mechanism works like this:
- viewport (x1y1, x2y2) is shown on screen
- c:geo loads caches within a significantly enlarged viewport of (a1b1, a2b2)
- subsequent requests are served from the cache, if
  - requested viewport is contained in (a1b1, a2b2)
  - and
    - last search returned less than 400 hits
    - or viewport is unchanged
    - or zoom level is greater than 11 (GM) / 14 (OSM)

Several situations can lead to not getting new caches (beside connection errors):
- even the originally requested viewport has more than max returned caches (500 for gc.com, IIRC)
- the original viewport is covered, but not the extended

One idea:
- after loading caches, calculate the min(x), max(x), min(y) and max(y) values of returned caches and waypoints (which lead to the minimum rectangle covering all returned geoitems) => (c1d1, c2d2)
- do a new search, if the requested viewport is not fully included in (c1d1, c2d2)
- maybe this should be combined with a limit - reduce (a1b1, a2b2) to (c1d1, c2d2) only if at least 400 caches have been returned?

This would lead to more accuracy potentially, but also to more calls to gc.com search, especially when moving the map around.

Maybe there's even a possibility to extend (c1d1, c2d2) if the new results don't lead to new caches.

What do you think?

**Comment by Lineflyer** [comment]
> 
    after loading caches, calculate the min(x), max(x), min(y) and max(y) values of returned caches and waypoints (which lead to the minimum rectangle covering all returned geoitems) => (c1d1, c2d2)
    do a new search, if the requested viewport is not fully included in (c1d1, c2d2)
    maybe this should be combined with a limit - reduce (a1b1, a2b2) to (c1d1, c2d2) only if at least 400 caches have been returned?

Yes, combined with the limit it makes sense IMHO.
Because if we initially request a viewport and receive less than maximum requested amount of caches we can consider the viewport covered in the cache.

Do we have to consider any overall size limit of our internal cache?

**Comment by moving-bits** [comment]
@Lineflyer 
I merged a PR implementing the proposal above. Can you test the resulting behavior?

**Comment by Lineflyer** [comment]
@moving-bits 
Will do

**Comment by Lineflyer** [comment]
I tested a bit with 2020.05.25-NB-c1460e3 using OSM offline maps:

- Zoom out on Düsseldorf city to reach maximum caches returned 
- Zoom in 
- Scroll to neighbor city Neuss (was part of viewport in step 1 but no caches shown due to max caches reached)

Result:
- Not all caches are shown for Neuss, no loading of new caches
- Only leaving and reentering the live map will load the caches

See video:
https://files.droescher.eu/public/20200525_091624.mp4

@moving-bits 
If you need debug logs or something, let me know. I will then repeat it.

**Comment by moving-bits** [comment]
Thanks for your feedback. Cool to have a video of it! 👍 

Hm, what's different between Düsseldorf/Neuss and Berlin (which is the city I had tested it with)...?

Anyway, tested it again, this time with Düsseldorf/Neuss - and it worked. Strange.

Also tested it with another city, Cologne in this case, and moving to south/west - worked as well.

![image](https://user-images.githubusercontent.com/3754370/82838208-c3dfdf80-9ecb-11ea-9e5d-e207fa15921b.png)

![image](https://user-images.githubusercontent.com/3754370/82838254-e70a8f00-9ecb-11ea-80a1-79e1ad8d1c6a.png)

Are you sure you had the most recent nightly in use?

Technical side: Reloading is triggered when the most recent search had returned at least 400 caches + when you have left the original bounding box (of the caches loaded) in at least one place.

During development I had added a red rectangle to visualize the current bounding box, which made it pretty easy to see when the bounding box changed due to newly loaded caches. Would it help for testing if I put that into the version?

**Comment by Lineflyer** [comment]
Looking at your screenshots the difference in your test is, that you did not increase the zoom between the steps:
- Load Düsseldorf at low zoom
- Zoom in a bit
- Scroll to Neuss

**Comment by moving-bits** [comment]
That's what I had done in my first step (just without screenshots). Redid it:

![image](https://user-images.githubusercontent.com/3754370/82839511-ee339c00-9ecf-11ea-84ef-89f462cf20b7.png)

![image](https://user-images.githubusercontent.com/3754370/82839529-f7246d80-9ecf-11ea-9139-6231bacdc8fc.png)

**Comment by moving-bits** [comment]
I've added the red bounding box, see how it enlarges from step to step while moving to south/west:

![image](https://user-images.githubusercontent.com/3754370/82842441-cb0dea00-9ed9-11ea-9cbf-e1fa61ace539.png)

![image](https://user-images.githubusercontent.com/3754370/82842448-d52fe880-9ed9-11ea-953a-bc32ca50196a.png)

![image](https://user-images.githubusercontent.com/3754370/82842454-dbbe6000-9ed9-11ea-80d8-ad18e766eee7.png)

![image](https://user-images.githubusercontent.com/3754370/82842459-e2e56e00-9ed9-11ea-8d82-f56e28fc3420.png)

![image](https://user-images.githubusercontent.com/3754370/82842543-440d4180-9eda-11ea-98b0-13b1f47e6705.png)

(Don't be confused by some caches being outside the box - that are those caches already having been loaded due to me moving the viewport to be able to make a screenshot of the red line.)

(This is GMv2 only, as the bounding box was easier to implement there. But the caching mechanism is the same, there should not be any differences.)

**Comment by Lineflyer** [comment]
Strange. Are you sure you have the same code base as it is merged to master?
Is there any memory of the viewports between sessions, or is it enough to force stop c:geo and have clean test conditions?
Because my video seems like a kind of evidence, that the problem is still there.

**Comment by moving-bits** [comment]
The screenshots above have all been made with today's master, so yes, should be the identical code base.

There is no viewport saved between sessions, AFAIK - hard-closing c:geo should be enough to restart with a fresh viewport.

I've added a small PR #8381 to bring you the red bounding box for testing. It's displayed automatically if you activate debug mode in c:geo (and use GMv2 map).

**Comment by Lineflyer** [comment]
@moving-bits 
I started a first look into the bounding box debug and have an idea about the difference:
I have multiple connectors active. Is the box calculation limited to GC or maybe also including other connectors?

**Comment by moving-bits** [comment]
The result is being calculated from the combined result. I can imagine two effects:
- bounding box being too large, when non GC caches are farther away than the farthest GC caches
- limit for reloading (400) reached earlier

**Comment by Lineflyer** [comment]
That could then indeed be the difference between our experiences. Can you try with enabled OC.de on the Düsseldorf/Neuss scenario?

**Comment by moving-bits** [comment]
Well, we may have caught it now... I created a new PR which calculates the caching viewport by only regarding gc.com caches, and only those which are not yet stored locally.

![image](https://user-images.githubusercontent.com/3754370/83063255-1fdb6d00-a060-11ea-8e44-9f8d2474c380.png)

![image](https://user-images.githubusercontent.com/3754370/83063285-27027b00-a060-11ea-93bf-8664a1dea639.png)

![image](https://user-images.githubusercontent.com/3754370/83063341-426d8600-a060-11ea-869d-8b90459c3060.png)

**Comment by Lineflyer** [comment]
Need to take a better look on it, but first test show, that (using the latest master incl. the change to consider only GC caches) now lead to situations, when sscrolling I got OC, EC, etc. caches loading  but GC still missing in that areas scrolled throug in a rather large extend.

**Comment by moving-bits** [comment]
Is this a "circle vs. square" thing?
gc.com delivers a circle of found caches, but the bounding box is calculated as a square, so the four corners are areas not covered by the search result.

**Comment by Lineflyer** [comment]
I am not sure what I tested on the first try, but I tried again today I now it looks pretty good (maybe I was still on a wrong version).
I tried with only GC and also with several connectors used and it looks good.

Naturally it will lead to some more requests to the server, but only on user interaction which should be fine. The original point from GS was, that requests happen without user interaction (probably due to the rotation) and that stays fixed.

So I guess we can call this one fixed until we learn otherwise. Thanks @moving-bits for your constant support in this.
BTW: We agreed to leave the debug function in case we still get complaints of missing caches to be investigated.
